README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->
beats
=====

Overview
--------

`beats` reads, visualizes, and process ECG data from UFI devices.

Installation
------------

``` r
# Install the development version from GitHub:
# install.packages("devtools")
devtools::install_github("FlukeAndFeather/beats")
```

Usage
-----

This example reads a binary UBE file and launches a Shiny app for finding heart beats.

``` r
# File path to sample data
fp <- system.file("extdata", "max_ecg_190826b.ube", package = "beats")
# Read ube file
ecg_data <- read_ube(fp)
# Launch GUI
heartbeats <- find_hr(ecg_data)
# In this heart rate profile, zoom into 12:53:20 - 12:54:25 to see actual beats
```

After identifying heart beats in ECG profile, you can create a GIF showing them in real time.

``` r
animate_hr(ecg_bw190918_62R, beats_bw190918_62R, which_beats = 44:50, big = TRUE)
```

![Example of heart rate animation](man/figures/heartbeats.gif)


NEWS.md
# beats 0.1.1

* Initial version
* Added raw, imported, and processed data
* Added functions for importing (`read_ube()`) and visualizing (`plot_ecg()`) heart rate data
* Add a function for processing data (`find_hr()`)
* Add a function for creating heart rate animations (`animate_hr()`)


tests/testthat.R
library(testthat)
library(beats)

test_check("beats")


tests/shinytest/app.R
shiny::shinyApp(ui = beats:::hr_ui,
                server = beats:::hr_server(beats::ecg_bw190918_62R),
                options = list(shiny.testmode = TRUE))


tests/testthat/test-read_ube.R
test_that("reading sample ECG data returns correct values", {
  # Read UBE data
  fp <- system.file("extdata", "max_ecg_190826b.ube", package = "beats")
  ecg_data <- read_ube(fp)
  ecg_head <- head(ecg_data)

  # Expected values
  expected_timestamp <- c("20-08-26 12:35:08.000", "20-08-26 12:35:08.009", "20-08-26 12:35:08.019", "20-08-26 12:35:08.029", "20-08-26 12:35:08.039", "20-08-26 12:35:08.049"
  )
  expected_ecg <- c(1418, 1418, 1476, 1382, 1467, 1424)

  # Tests
  expect_equal(format(ecg_head$timestamp, "%y-%m-%d %H:%M:%OS3"), expected_timestamp)
  expect_equal(ecg_head$ecg, expected_ecg)
})


R/data.R
#' ECG of a blue whale
#'
#' A dataset containing the ECG record of a blue whale, collected 2019-09-18
#' off the coast of Big Sur, CA under NMFS permit no. 16111.
#'
#' @format A data frame with 234,001 rows and two variables:
#' \describe{
#'   \item{timestamp}{time of record, in POSIXct format}
#'   \item{ecg}{value from ECG recorder in engineering units, ranges 0-4095}
#' }
#'
#' @references Goldbogen, J. A., Cade, D. E., Calambokidis, J., Czapanskiy, M. F., Fahlbusch, J., Friedlaender, A. S., … Ponganis, P. J. (2019). Extreme bradycardia and tachycardia in the world’s largest animal. *Proceedings of the National Academy of Sciences, 116*(50), 25329–25332. doi: 10.1073/pnas.1914273116
"ecg_bw190918_62R"

#' Heart rate of a blue whale
#'
#' A dataset containing the heart rate record of a blue whale, collected
#' 2019-09-18 off the coast of Big Sur, CA under NMFS permit no. 16111.
#'
#' @format A data frame with 59 rows and five variables:
#' \describe{
#'   \item{timestamp}{time of heart beat, in POSIXct format}
#'   \item{ecg}{value from ECG recorder at peak voltage (see \link{ecg_bw190918_62R})}
#'   \item{period_s}{period in seconds to next heart beat. NA if followed by a gap.}
#'   \item{freq_hz}{frequency of heart beat in Hz (inverse of period_s)}
#'   \item{freq_bpm}{frequency of heart beat in beats-per-minute}
#' }
#'
#' @references Goldbogen, J. A., Cade, D. E., Calambokidis, J., Czapanskiy, M. F., Fahlbusch, J., Friedlaender, A. S., … Ponganis, P. J. (2019). Extreme bradycardia and tachycardia in the world’s largest animal. *Proceedings of the National Academy of Sciences, 116*(50), 25329–25332. doi: 10.1073/pnas.1914273116
"beats_bw190918_62R"


R/plot_ecg.R
#' Plot a heartrate profile
#'
#' \code{plot_ecg} plots a heat rate profile.
#'
#' @param data A data.frame with columns timestamp (POSIXct) and ecg (numeric)
#' @param max_points Maximum number of points to decimate \code{data} (10,000
#' by default)
#' @param interactive Should the plot be interactive? FALSE by default.
#' Requires package "plotly".
#' @return A ggplot object if not interactive, a plotly object otherwise.
#' @export
plot_ecg <- function(data, max_points = 1e4, interactive = FALSE) {
  # Check inputs
  stopifnot(is.data.frame(data))
  stopifnot(c("timestamp", "ecg") %in% data)
  stopifnot(is.numeric(max_points))
  stopifnot(length(max_points) == 1)
  stopifnot(max_points > 0)
  max_points <- as.integer(max_points)

  if (interactive && !requireNamespace("plotly", quietly = TRUE)) {
    stop("Package \"plotly\" must be installed for interactive plots")
  }

  # Plot heart rate profile
  p <- data %>%
    dplyr::slice(seq(1, nrow(data), length.out = max_points)) %>%
    ggplot2::ggplot(ggplot2::aes(timestamp, ecg)) +
    ggplot2::geom_line(size = 0.5) +
    ggplot2::scale_x_datetime(date_labels = "%H:%M") +
    ggplot2::theme_classic()

  if (interactive) {
    plotly::ggplotly(p)
  } else {
    p
  }
}


R/read_ube.R
#' Read data from a UBE file
#'
#' \code{read_ube} parses a binary UBE file, returning a data frame with ECG
#' records and timestamps
#'
#' @param ube_path Path to UBE file.
#' @return A data.frame with columns timestamp (POSIXct) and ecg (numeric)
#' @examples
#' # File path to sample data
#' fp <- system.file("extdata", "max_ecg_190826b.ube", package = "beats")
#' # Read ube file
#' ecg_data <- read_ube(fp)
#' @export
read_ube <- function(ube_path) {
  # Read the raw data
  ube_raw <- readr::read_file_raw(ube_path)

  # First 32 bytes are the download timestamp
  dl_time <- rawToChar(ube_raw[1:32], multiple = TRUE) %>%
    paste(collapse = "") %>%
    lubridate::mdy_hms(tz = "UTC")

  # Next 5 bytes are the time of recording beginning
  # Not sure what bytes 38-40 are
  # Year doesn't seem to be recorded, so assume year is the same as now
  mdhms <- as.numeric(ube_raw[33:37])
  record_start <- ISOdatetime(lubridate::year(lubridate::now()), # year
                              mdhms[1],                          # month
                              mdhms[2],                          # day
                              mdhms[3],                          # hour
                              mdhms[4],                          # minute
                              mdhms[5],                          # second
                              tz = "UTC")

  # Remainder is data
  # Each record is 2 bytes, of the format 0xCddd where C is the channel (2 for
  # ECG) and ddd is the value

  # Remove the header
  data_raw <- ube_raw[-(1:40)]
  # ecg_channel is a mask for checking if a record is ECG data
  ecg_channel <- 0x2F
  mode(ecg_channel) <- "raw"
  # Since each record is two bytes, the output is at largest half the length of
  # the raw data
  ecg_data <- integer(length(data_raw) / 2)

  i <- 1
  for (raw_i in seq(1, length(data_raw), by = 2)) {
    # Check if a record is ecg data (i.e. starts with 0x2), otherwise skip it
    if ((ecg_channel | data_raw[raw_i]) != ecg_channel)
      next
    # This math gets the value of a nibble plus a byte
    num <- as.numeric(data_raw[raw_i:(raw_i+1)])
    ecg_data[i] <- (num[1] - 0x20) * 0x100 + num[2]
    i <- i + 1
  }

  # Trim the data to valid values
  ecg_data <- ecg_data[1:(i - 1)]

  # The timestamp of the ECG values assumes 100 Hz
  # In the future, the sampling rate could be read from the header
  ecg_time <- record_start + (seq_along(ecg_data) - 1) / 100

  # Create a result table and set an attribute for the creation time.
  result <- data.frame(timestamp = ecg_time,
                       ecg = ecg_data)
  attr(result, "created") <- dl_time
  result
}


R/shiny_app.R
# Also see shiny_helpers.R

# UI
hr_ui <- shiny::fillPage(
  shiny::sidebarLayout(
    shiny::sidebarPanel(shiny::radioButtons("mode",
                                            "Interaction mode",
                                            list(`Add heart beat` = "add_beat",
                                                 `Clear heart beat` = "clear_beat",
                                                 `Add gap` = "add_gap",
                                                 `Clear gap` = "clear_gap",
                                                 `Set beat threshold` = "beat_thr")),
                        shiny::downloadButton("output_csv", "Download heart beats (.csv)"),
                        shiny::actionButton("done", "Finish and return data.frame")),
    shiny::mainPanel(
      shiny::plotOutput("data_plot",
                        brush = "data_brush",
                        height = "100px"),
      shiny::plotOutput("deploy_plot",
                        brush = "deploy_brush",
                        height = "100px"),
      shiny::plotOutput("detail_plot",
                        click = "detail_click",
                        brush = "detail_brush",
                        height = "400px"))
  )
)


# SERVER
hr_server <- function(data) {
  shiny::shinyServer(function(input, output, session) {
    values <- shiny::reactiveValues(ecg_data = data)

    # When brushing profiles, remove downstream brushes and update data
    shiny::observeEvent(input$data_brush, {
      session$resetBrush("deploy_brush")
      session$resetBrush("detail_brush")
      values$ecg_deploy <- subdata(shiny::isolate(values$ecg_data),
                                   input$data_brush)
    })
    shiny::observeEvent(input$deploy_brush, {
      session$resetBrush("detail_brush")
      values$ecg_detail <- subdata(shiny::isolate(values$ecg_deploy),
                                    input$deploy_brush)
    })

    # Handle clicking and brushing on detail plot
    shiny::observeEvent(input$detail_click, {
      result <- handle_detail_click(shiny::isolate(values$ecg_detail),
                                    input$detail_click,
                                    shiny::isolate(input$mode),
                                    shiny::isolate(values$heart_beats),
                                    shiny::isolate(values$ecg_gaps))
      # Update heart_beats and ecg_gaps
      values$heart_beats <- result$heart_beats
      values$ecg_gaps <- result$ecg_gaps
    })
    shiny::observeEvent(input$detail_brush, {
      result <- handle_detail_brush(shiny::isolate(values$ecg_detail),
                                    input$detail_brush,
                                    shiny::isolate(input$mode),
                                    shiny::isolate(values$heart_beats),
                                    shiny::isolate(values$ecg_gaps))
      # Update heart_beats and ecg_gaps
      values$heart_beats <- result$heart_beats
      values$ecg_gaps <- result$ecg_gaps
    })

    # Data plot (profile of all data)
    output$data_plot <- shiny::renderPlot(plot_profile(values$ecg_data))
    # Deploy plot (profile of deployment data)
    output$deploy_plot <- shiny::renderPlot(plot_profile(values$ecg_deploy,
                                                         values$heart_beats,
                                                         values$ecg_gaps))
    # Detail plot (zoomed in data)
    output$detail_plot <- shiny::renderPlot(plot_profile(values$ecg_detail,
                                                         values$heart_beats,
                                                         values$ecg_gaps,
                                                         detail = TRUE))

    # Download CSV
    output$output_csv <- shiny::downloadHandler(
      filename = function() {
        format(lubridate::now(), "ecg_%y%m%d%H%M%S.csv")
      },
      content = function(file) {
        readr::write_csv(prepare_csv(values$heart_beats,
                                     values$ecg_gaps),
                         file)
      })

    # Finish and return beats
    shiny::observeEvent(input$done, {
      shiny::stopApp(returnValue = prepare_beats(values$heart_beats,
                                                 values$ecg_gaps))
    })
  })
}


R/animate_hr.R
#' Animate ECG data and heart beats
#'
#' \code{animate_hr} produces a GIF of heart beats in (close to) real time.
#'
#' @param ecg A data.frame with columns timestamp (POSIXct) and ecg (numeric)
#' (see \code{\link{read_ube}})
#' @param beats A data.frame with columns timestamp (POSIXct), ecg, period_s,
#' freq_hz, and freq_bpm (see \code{\link{find_hr}})
#' @param which_beats Indices of the heart beats (in param \code{beats}) to be
#' animated. If NULL (default), all beats will be animated.
#' @param big If FALSE (default) then throw an error if the animation will last
#' more than 20s. This prevents R from crashing if it tries to animate too
#' much.
#' @return A \code{\link[gganimate]{gif_image}} object
#'
#' @examples
#' \dontrun{
#'   animate_hr(ecg_bw190918_62R, beats_bw190918_62R, which_beats = 44:50, big = TRUE)
#' }
#'
#' @md
#' @export
animate_hr <- function(ecg, beats, which_beats = NULL, big = FALSE) {
  # Check if required packages are installed
  required_packages <- c("gganimate", "gifski", "png")
  if (any(!required_packages %in% utils::installed.packages())) {
    stop("To use animate_hr install gganimate, gifski, and png")
  }

  # Check input data
  if (!all(c("timestamp", "ecg") %in% colnames(ecg))) {
    stop("Columns of table ecg must include 'timestamp' and 'ecg'.")
  }
  if (!all(c("timestamp", "ecg", "freq_bpm") %in% colnames(beats))) {
    stop("Columns of table beats must include 'timestamp', 'ecg', and 'freq_bpm'.")
  }
  if (!is.null(which_beats)) {
    if (!is.numeric(which_beats))
      stop("which_beats must be numeric")
    if (any(!which_beats %in% seq_along(beats$timestamp)))
      stop("which_beats must be valid indices of beats")
  }
  if (!is.logical(big) || length(big) != 1)
    stop("big must be a logical of length 1")

  # If which_beats is NULL, animate all heart beats
  if (is.null(which_beats)) {
    which_beats = seq_along(beats$timestamp)
  }

  # gganimate requires data.frames, not tibbles
  beats2 <- dplyr::slice(beats, which_beats) %>%
    dplyr::mutate(bpm_lbl = sprintf("%0.2f", freq_bpm)) %>%
    as.data.frame()

  # Add a 0.5 buffer on either side of the heart beats for ecg data, expanded
  # to the nearest second. gganimate requires whole seconds.
  buffer <- .5 # seconds
  ecg_start <- lubridate::floor_date(beats2$timestamp[1] - buffer,
                                     unit = "second")
  ecg_end <- lubridate::ceiling_date(beats2$timestamp[nrow(beats2)] + buffer,
                                     unit = "second")
  ecg2 <- dplyr::filter(ecg,
                        dplyr::between(timestamp, ecg_start, ecg_end)) %>%
    as.data.frame()

  # Calculate duration of animation, fps, and nframes
  dur_s <- as.numeric(max(ecg2$timestamp) - min(ecg2$timestamp), unit = "secs")
  if (dur_s %% 1 != 0)
    stop("Something went wrong, dur_s should be a whole number")
  if (!big && dur_s > 20) {
    stop("For large animations (>20s) big must be TRUE")
  }
  fps <- 10
  nframes <- dur_s * fps

  # Animate heart rate
  hr_anim <- ggplot2::ggplot(ecg2, ggplot2::aes(timestamp, ecg)) +
    ggplot2::geom_line(size = 0.1) +
    ggplot2::geom_point(ggplot2::aes(group = seq_along(timestamp)),
                        data = beats2,
                        shape = 3,
                        color = "red") +
    ggplot2::geom_text(ggplot2::aes(group = seq_along(timestamp),
                                    label = bpm_lbl),
                       data = beats2,
                       nudge_y = 100) +
    ggplot2::scale_x_datetime("", date_labels = "%H:%M:%S", timezone = Sys.timezone()) +
    ggplot2::ylim(0, 4095) +
    ggplot2::theme_minimal() +
    gganimate::transition_reveal(timestamp)
  gganimate::animate(hr_anim, nframes, fps, dur_s)
}


R/shiny_helpers.R
# Plot an ECG profile
plot_profile <- function(data, beats = NULL, gaps = NULL, detail = FALSE) {
  shiny::req(data)

  # Maximum 10,000 points
  data_sparse <- data %>%
    dplyr::slice(seq(1, nrow(data), length.out = 1e4))

  # Date breaks
  date_rng <- max(data_sparse$timestamp) - min(data_sparse$timestamp)
  offset <- 0.05
  date_breaks <- seq(min(data_sparse$timestamp) + offset * date_rng,
                     max(data_sparse$timestamp) - offset * date_rng,
                     length.out = 5)

  # Profile plots show dates in ymd hms, detail is just hms
  tz <- lubridate::tz(data_sparse$timestamp)
  if (tz == "")
    tz <- Sys.timezone()
  if (detail) {
    lbl <- "%H:%M:%S"
  } else {
    lbl <- "%y-%m-%d %H:%M"
  }
  x_scale <- ggplot2::scale_x_datetime(date_labels = lbl,
                                       timezone = tz,
                                       breaks = date_breaks)

  # Plot
  p <- ggplot2::ggplot(data_sparse, ggplot2::aes(timestamp, ecg)) +
    ggplot2::geom_line(size = 0.1) +
    x_scale +
    ggplot2::ylim(0, 4095) +
    ggplot2::theme_classic() +
    ggplot2::theme(axis.title = ggplot2::element_blank())

  # Add points for heart beats
  if (!is.null(beats)) {
    beats_visible <- dplyr::filter(beats,
                                   dplyr::between(timestamp,
                                                  dplyr::first(data$timestamp),
                                                  dplyr::last(data$timestamp)))
    p <- p + ggplot2::geom_point(data = beats_visible,
                                 shape = 3,
                                 color = "red")
  }

  # Add rects for gaps
  if (!is.null(gaps)) {
    gaps_visible <- dplyr::filter(gaps,
                                  timestamp_end > min(data$timestamp),
                                  timestamp_begin < max(data$timestamp))
    if (nrow(gaps_visible) > 0) {
      gaps_visible <- gaps_visible %>%
        dplyr::mutate(gap_xmin = pmax(min(data$timestamp), timestamp_begin),
                      gap_xmax = pmin(max(data$timestamp), timestamp_end))
      p <- p + ggplot2::geom_rect(ggplot2::aes(xmin = gap_xmin,
                                               xmax = gap_xmax,
                                               ymin = 0,
                                               ymax = 4095),
                                  gaps_visible,
                                  inherit.aes = FALSE,
                                  fill = "blue",
                                  alpha = 0.1)
    }
  }

  p
}

# Handle clicking on detail plot
handle_detail_click <- function(data, click, mode, beats, gaps) {
  # Wait for data and a click
  shiny::req(data, click)
  switch (mode,
          add_beat = add_beat(data, click, beats, gaps),
          clear_beat = clear_beat(data, click, beats, gaps),
          clear_gap = clear_gap(data, click, beats, gaps),
          beat_thr = beat_thr(data, click, beats, gaps),
          list(heart_beats = beats,
               ecg_gaps = gaps)
  )
}

# Handle brushing the detail plot
handle_detail_brush <- function(data, brush, mode, beats, gaps) {
  # Wait for data and a brush
  shiny::req(data, brush)
  switch (mode,
          add_gap = add_gap(data, brush, beats, gaps),
          list(heart_beats = beats,
               ecg_gaps = gaps)
  )
}

# Add heart beat
add_beat <- function(data, click, beats, gaps) {
  nearest_ecg <- which.min(abs(click$x - as.numeric(data$timestamp)))
  # Look for a peak within 0.5s
  search_range <- dplyr::filter(data,
                                timestamp > data$timestamp[nearest_ecg] - 0.5,
                                timestamp < data$timestamp[nearest_ecg] + 0.5)
  nearest_peak <- dplyr::filter(search_range, ecg == max(ecg)) %>%
    dplyr::slice(1)
  # Add peak to beats
  new_beats <- rbind(nearest_peak, beats) %>%
    # Sort chronologically
    dplyr::arrange(timestamp) %>%
    # Remove duplicates
    dplyr::distinct()

  # Return beats and gaps
  list(heart_beats = new_beats,
       ecg_gaps = gaps)
}

# Clear heart beat
clear_beat <- function(data, click, beats, gaps) {
  new_beats <- beats

  # Filter beats to visible in details
  beats_visible <- dplyr::filter(beats,
                                 dplyr::between(timestamp,
                                                dplyr::first(data$timestamp),
                                                dplyr::last(data$timestamp)))

  # If any beats are visible...
  if (nrow(beats_visible) > 0) {
    nearest_beat <- which.min(abs(click$x - as.numeric(beats_visible$timestamp)))
    drop_timestamp <- beats_visible$timestamp[nearest_beat]
    # ... drop the closest one within 0.5s
    if (abs(click$x - as.numeric(drop_timestamp)) < 0.5) {
      new_beats <- dplyr::filter(beats, timestamp != drop_timestamp)
    }
  }

  # Return beats and gaps
  list(heart_beats = new_beats,
       ecg_gaps = gaps)
}

# Add a gap
add_gap <- function(data, brush, beats, gaps) {
  # Nearest points to brush borders define the new gap
  nearest_ecg <- integer(2)
  nearest_ecg[1] <- which.min(abs(brush$xmin - as.numeric(data$timestamp)))
  nearest_ecg[2] <- which.min(abs(brush$xmax - as.numeric(data$timestamp)))
  new_gap <- data.frame(timestamp_begin = data$timestamp[nearest_ecg[1]],
                        timestamp_end = data$timestamp[nearest_ecg[2]])

  # Add new_gap to gaps
  new_gaps <- rbind(gaps, new_gap) %>%
    dplyr::arrange(timestamp_begin)

  # Remove beats under gap
  new_beats <- beats
  if (!is.null(beats) && nrow(beats) > 0) {
    new_beats <- dplyr::filter(beats,
                               !dplyr::between(timestamp,
                                               new_gap$timestamp_begin,
                                               new_gap$timestamp_end))
  }

  # Merge overlapping gaps
  last_end <- max(c(new_gaps$timestamp_end, new_gaps$timestamp_end)) + 1
  overlapping <- which(new_gaps$timestamp_end > dplyr::lead(new_gaps$timestamp_begin))
  while (length(overlapping) > 0) {
    i <- overlapping[1]
    new_gaps$timestamp_end[i] <- new_gaps$timestamp_end[i + 1]
    new_gaps <- new_gaps[-(i + 1), ]
    overlapping <- which(new_gaps$timestamp_end > dplyr::lead(new_gaps$timestamp_begin))
  }

  list(heart_beats = new_beats,
       ecg_gaps = new_gaps)
}

# Clear a gap
clear_gap <- function(data, click, beats, gaps) {
  new_gaps <- gaps
  # Look for a gap under the click and remove it
  gap <- which(click$x > gaps$timestamp_begin & click$x < gaps$timestamp_end)
  if (length(gap) == 1)
    new_gaps <- dplyr::slice(gaps, -gap)

  list(heart_beats = beats,
       ecg_gaps = new_gaps)
}

# Set beat threshold
beat_thr <- function(data, click, beats, gaps) {
  # Look for peaks in 1s windows
  window_size <- 100
  local_peaks <- zoo::rollapply(data$ecg,
                                window_size,
                                function(x) which.max(x) == window_size / 2,
                                partial = TRUE)

  # Remove local peaks under threshold
  local_peaks[data$ecg < click$y] <- FALSE

  # Add new beats
  new_beats <- rbind(beats, data[local_peaks,]) %>%
    dplyr::arrange(timestamp) %>%
    dplyr::distinct()

  # Remove new beats if under gap
  # Remove beats under gap
  if (!is.null(gaps)) {
    visible_gaps <- dplyr::filter(gaps,
                                  timestamp_end >= min(data$timestamp),
                                  timestamp_begin <= max(data$timestamp))
    if (nrow(visible_gaps) > 0) {
      # utility function, checks if a timestamp falls in any visible gap
      is_under_gap <- function(t) {
        purrr::map_lgl(t, function(t) {
          any(t >= visible_gaps$timestamp_begin &
                t <= visible_gaps$timestamp_end)
        })
      }
      # data frame of beats under gaps
      gap_beats <- dplyr::mutate(new_beats,
                                 under_gap = is_under_gap(timestamp)) %>%
        dplyr::filter(under_gap)
      # anti join gap beats out of new beats
      new_beats <- dplyr::anti_join(new_beats, gap_beats, by = "timestamp")
    }
  }

  list(heart_beats = new_beats,
       ecg_gaps = gaps)
}

# Update plot data
subdata <- function(data, brush) {
  shiny::req(data, brush)
  dplyr::filter(data,
                timestamp >= brush$xmin,
                timestamp <= brush$xmax)
}

# prepare_beats
# Input: beats, gaps
# Output: data frame
prepare_beats <- function(beats, gaps) {
  result <- beats %>%
    dplyr::mutate(period_s = as.numeric(dplyr::lead(timestamp) - timestamp,
                                        unit = "secs"),
                  freq_hz = 1 / period_s,
                  freq_bpm = freq_hz * 60)
  # Find the heart beat before each gap and remove period/freq
  gap_beats <- unique(findInterval(gaps$timestamp_begin, result$timestamp))
  gap_beats <- gap_beats[gap_beats > 0]
  # Keep timestamp and ecg (cols 1 and 2) but set everything else to NA
  result[gap_beats, -(1:2)] <- NA

  # Return result
  result
}
# prepare_rdata
# Input: ecg_deploy, beats, gaps
# Output:
prepare_rds <- function(data, deploy, detail, beats, gaps) {
  list(data = data,
       deploy = deploy,
       detail = detail,
       beats = beats,
       gaps = gaps,
       output = prepare_beats(beats, gaps))
}
prepare_csv <- function(beats, gaps) {
  prepare_beats(beats, gaps)
}


R/find_hr.R
#' Find heart rate from ECG data
#'
#' \code{find_hr} is an interactive tool for processing ECG data to get heart
#' rate profiles. Use it to identify individual heart beats and gaps in ECG
#' the ECG data.
#'
#' @param data A data.frame with columns timestamp (POSIXct) and ecg (numeric)
#' @return A data.frame with columns
#'   * timestamp (time of heart beat, POSIXct)
#'   * ecg (raw value of ECG peak)
#'   * period_s (period in seconds to next heart beat)
#'   * freq_hz (heart rate in Hz)
#'   * freq_bpm (heart rate in beats per minute)
#' @details The GUI is divided into a side panel with controls and a main panel
#'   with plots. The side panel has controls for choosing the interaction mode,
#'   exporting results, and exiting. The main panel shows up to three plots for
#'   viewing and interacting with the data. As you interact with the plots,
#'   more detailed views appear.
#'
#'   1. The top plot shows the entire ECG profile, including data before and
#'     after deployment. Brush (click and drag) the part of the profile with
#'     the deployment itself.
#'   2. The middle plot is a profile of the deployment. Brush it to zoom in to
#'     shorter periods so you can see individual heart beats.
#'   3. The bottom plot shows data in the most detailed view. Clicking and
#'     brushing the data will add/clear heart beats and gaps.
#'
#'   There are five interaction modes. In "Add heart beat" mode, clicking will
#'   add a heart beat on the nearest peak (should be the QRS complex). "Clear
#'   heart beat" will remove the nearest beat. Use "Add gap" to brush areas
#'   with unusable data, to avoid artificially low heart rates. Overlapping gaps
#'   are merged automatically. In "Clear gap" mode, click on gaps to clear
#'   them. "Set beat threshold" adds heart beats in bulk. Clicking will add a
#'   heart beat to every peak higher than the click point. This will only add
#'   beats to the part of the data visible in the detail plot.
#'
#'   There are two buttons at the bottom of the side panel for when you're done
#'   using the tool. "Download heart beats" exports the heart rate data to CSV
#'   and "Finish and return data.frame" quits the tool and returns the result.
#'
#' @examples
#' \dontrun{
#' # File path to sample data
#' fp <- system.file("extdata", "max_ecg_190826b.ube", package = "beats")
#' # Read ube file
#' ecg_data <- read_ube(fp)
#' # Launch GUI
#' heartbeats <- find_hr(ecg_data)
#' # In this heart rate profile, zoom into 12:53:20 - 12:54:25 to see actual beats
#' }
#' @md
#' @export
find_hr <- function(data) {
  options(shiny.maxRequestSize = 100*1024^2)
  # hr_ui and hr_server are in shiny_app.R
  shiny::runApp(list(ui = hr_ui, server = hr_server(data)))
}
